File: Programmers/Level3/징검다리건너기.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

int solution(vector<int> stones, int k) {
	int answer = 200000001;

	for (int i = 0; i < stones.size()-k+1; i++) {
		int curMax = 0;
		int maxIndex = i;
		for (int j = i; j < i+k; j++) {
			if (curMax < stones[j]) {
				curMax = stones[j];
				maxIndex = j;
			}
		}
		if (i + 1 < maxIndex) i = maxIndex - 1;
		if(answer>curMax) answer = curMax;
	}
	return answer;
}

File: Programmers/Level3/징검다리건너기.java
import java.util.PriorityQueue;

class Node {
    int index;
    int value;
    Node(int index, int value) {
        this.index = index;
        this.value = value;
    }
}
class Solution {

    public int solution(int[] stones, int k) {
        PriorityQueue<Node> pq = new PriorityQueue<Node>(((o1, o2) -> {
            if(o1.value > o2.value )return -1;
            else if(o1.value < o2.value) return 1;
            else return 0;
        }));
        for(int i=0; i<k; i++) {
            pq.add(new Node(i, stones[i]));
        }
        int result = pq.peek().value;
        for(int i=k; i<stones.length; i++) {
            pq.add(new Node(i, stones[i]));
            while(!pq.isEmpty()) {
                if(pq.peek().index <= i-k) pq.poll();
                else break;
            }
            result = Math.min(result, pq.peek().value);
        }
        return result;
    }
}
